src/levanter/compat/fsspec_safetensor.py

```diff
@@ -0,0 +1,290 @@
+        self._fs = fs
+        import concurrent.futures
+
+        default_workers = min(64, max(8, (os.cpu_count() or 8) * 2))
+        max_workers = int(os.environ.get("LEVANTER_FSSPEC_MAX_WORKERS", default_workers))
+        self._executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
+
+    async def _cat_file(self, path: str, start: int | None = None, end: int | None = None, **kwargs) -> bytes:
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(
+            self._executor,
+            lambda: self._fs.cat_file(path, start=start, end=end, **kwargs),
+        )
+
+    async def _info(self, path, **kwargs):
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(
+            self._executor,
+            lambda: self._fs.info(path, **kwargs),
+        )
+
+    async def _size(self, path):
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(
+            self._executor,
+            lambda: self._fs.size(path),
+        )
+
+
+# See https://huggingface.co/docs/safetensors/en/index#format for format spec
+# It's pretty simple:
+# # - 8 bytes: little-endian uint64 header length N
+# # - N bytes: UTF-8 JSON header of shapes/dtypes/data offsets
+# # - remaining bytes: raw tensor data blobs
+async def _read_metadata_async(fs: AsyncFileSystem, path: str) -> Dict[str, TensorRecord]:
+    header_len_bytes = await fs._cat_file(path, start=0, end=8)
+    (header_len,) = struct.unpack("<Q", header_len_bytes)
+    metadata_bytes = await fs._cat_file(path, start=8, end=8 + header_len)
+    metadata = json.loads(metadata_bytes.decode("utf-8"))
+
+    tensors: Dict[str, TensorRecord] = {}
+    data_offset_base = 8 + header_len
+
+    for key, meta in metadata.items():
+        if key == "__metadata__":
+            continue
+        dtype_name: str = meta["dtype"]
+        dtype = _SAFETENSOR_DTYPE_MAP.get(dtype_name)
+        if dtype is None:
+            raise ValueError(f"Unsupported safetensors dtype: {dtype_name}")
+
+        rel_start, rel_end = meta["data_offsets"]
+        tensors[key] = TensorRecord(
+            key=key,
+            dtype=dtype,
+            shape=tuple(meta["shape"]),
+            file_path=path,
+            byte_start=data_offset_base + rel_start,
+            byte_end=data_offset_base + rel_end,
+        )
+
+    return tensors
+
+
+@dataclass(frozen=True)
+class ChunkSpec:
+    file_path: str
+    byte_start: int
+    byte_end: int
+    tensors: Tuple[TensorRecord, ...]
+
+    @property
+    def size(self) -> int:
+        return self.byte_end - self.byte_start
+
+
+def _build_chunks(tensors: Iterable[TensorRecord], chunk_limit: int) -> List[ChunkSpec]:
+    if chunk_limit <= 0:
+        raise ValueError("chunk_limit must be positive")
+
+    sorted_records = sorted(tensors, key=lambda t: (t.file_path, t.byte_start))
+    chunks: List[ChunkSpec] = []
+
+    current: List[TensorRecord] = []
```

Comment:
You know this already but we'll likely spend our time reading the giant tensors out as single chunks. Doesn't seem worth optimizing that this point though.